src/fs.rs

```diff
@@ -652,56 +639,51 @@ impl<S: 'static> StaticFiles<S> {
         self.default = Box::new(WrapHandler::new(handler));
         self
     }
+
+    fn try_handle(&self, req: &HttpRequest<S>) -> Result<AsyncResult<HttpResponse>, Error> {
+        let tail: String = req.match_info().query("tail")?;
+        let relpath = PathBuf::from_param(tail.trim_left_matches('/'))?;
+
+        // full filepath
+        let path = self.directory.join(&relpath).canonicalize()?;
+
+        if path.is_dir() {
+            if let Some(ref redir_index) = self.index {
+                // TODO: Don't redirect, just return the index content.
+                // TODO: It'd be nice if there were a good usable URL manipulation
+                // library
+                let mut new_path: String = req.path().to_owned();
+                if !new_path.ends_with('/') {
+                    new_path.push('/');
+                }
+                new_path.push_str(redir_index);
+                HttpResponse::Found()
+                    .header(header::LOCATION, new_path.as_str())
+                    .finish()
+                    .respond_to(&req)
+            } else if self.show_index {
+                let dir = Directory::new(self.directory.clone(), path);
+                Ok((*self.renderer)(&dir, &req)?.into())
+            } else {
+                Err(StaticFileError::IsDirectory.into())
+            }
+        } else {
+            NamedFile::open(path)?
+                .set_cpu_pool(self.cpu_pool.clone())
+                .respond_to(&req)?
+                .respond_to(&req)
+        }
+    }
 }
 
 impl<S: 'static> Handler<S> for StaticFiles<S> {
     type Result = Result<AsyncResult<HttpResponse>, Error>;
 
     fn handle(&self, req: &HttpRequest<S>) -> Self::Result {
-        if !self.accessible {
+        self.try_handle(req).or_else(|e| {
+            debug!("StaticFiles: Failed to handle {}: {}", req.path(), e);
```

Comment:
I'm not sure should we log each failure to handle request?

There potentially can be lots of ill-formed requests, the only suspicious case I'd suggest to cover by such trace is `canonicalize()` but other variants such as non-existing file or `StaticFileError::IsDirectory` is not really a failure, but expected outcome when user requests unknown resource.

Reply:
From the point of view of the library, I'd agree with you. Any errors logged here are expected behaviour. But as someone using the library, if I'm getting a 404 error, I'd like to know _why_. Is it because a file that's supposed to be there is missing? Is it because I forgot to set the index file? Is it because I don't have permission to open the file? That's why I put it at `debug!` level, to make it easier to debug things as a library user but so it wouldn't show normally. But if you'd rather I remove it then sure, I can do that.

Reply:
I certainly understand your reasoning, I'm not strongly against it as it is possible to remove traces using compile time features of `log` crate.

@fafhrd91 Should we have debug trace here? Originally there was no such trace

Reply:
Just to explain where I'm coming from, I'm used to using rails which logs everything under the sun at debug level (including all web requests, all parameters passed, which controller action handles the request, all database requests, all templates rendered and stack traces on any error caught). It's a lot of stuff but it's also pretty useful. But it's definitely up to you guys what you want here.

Reply:
debug is fine